### ISSUE TITLE
Support for HuggingFace tokenizer

### ISSUE THREAD

**Issue by zhengyang-wang** [facebookresearch/lingua#32]
Is there a plan to support HuggingFace tokenizer?

**Comment by zhengyang-wang** [comment]
```
class HuggingFaceTokenizer(Tokenizer):
    def __init__(self, model_path):
        self.hf_tokenizer = HFTokenizer.from_pretrained(model_path)

        logger.info(f"Reloaded HFTokenizer model from {model_path}")

        self.bos_id: int = self.hf_tokenizer.encode(self.hf_tokenizer.bos_token)[0]
        self.eos_id: int = self.hf_tokenizer.encode(self.hf_tokenizer.eos_token)[0]
        self.n_words: int = max(self.hf_tokenizer.vocab.values())+1

    def encode(self, tokens, add_bos, add_eos):
        assert type(tokens) is str
        return (
            [self.bos_id] * add_bos
            + self.hf_tokenizer.encode(tokens)
            + [self.eos_id] * add_eos
        )

    def decode(self, tokens):
        return self.hf_tokenizer.decode(tokens)

    def get_token_offsets(self, text, tokens=None):
        # This implementation comes from GPT-4o
        # Use the encode_plus method to get token offsets
        encoded_output = self.hf_tokenizer.encode_plus(
            text,
            return_offsets_mapping=True,  # To get the offsets
            add_special_tokens=False  # Avoid adding special tokens like BOS/EOS in this method
        )
        
        # Extract the offsets (a list of tuples with start and end positions)
        offsets = encoded_output["offsets_mapping"]
        
        # Extract the tokenized substrings
        substrs = [text[start:end] for start, end in offsets]

        # Extract the starting positions from the offset tuples
        start_offsets = [start for start, _ in offsets]

        return substrs, start_offsets
```

**Comment by lopezpaz** [comment]
Feel free to re-open if this is not already addressed in #27.

**Comment by Hannibal046** [comment]
@zhengyang-wang

Hi, thank you so much for providing the code! I noticed something and was wondering if there's a missing argument for the Hugging Face tokenizer's encode method. Specifically:
```python
def encode(self, tokens, add_bos, add_eos):
    assert type(tokens) is str
    return (
        [self.bos_id] * add_bos
        + self.hf_tokenizer.encode(tokens, add_special_tokens=False)
        + [self.eos_id] * add_eos
    )
```
